File: 20230712_Wednesday/20230712.rb
# Leetcode: 802. Find Eventual Safe States.
# https://leetcode.com/problems/find-eventual-safe-states/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 209 ms, faster than 50.00% of Ruby online submissions for Find
# Eventual Safe States.
# Memory Usage: 220.6 MB, less than 50.00% of Ruby online submissions for Find
# Eventual Safe States.
# 2023.07.12 Daily Challenge.
# @param {Integer[][]} graph
# @return {Integer[]}
def eventual_safe_nodes(graph)
  @g = graph
  @n = graph.size
  n = @n
  @v = Array.new(n,false)  # Visted.
  @ans = []
  @type = Array.new(n,0) # 0 - no info, 1 - safe, 2 unsafe
  @n.times do |i|
    dfs(i) if ! @v[i]
  end
  @ans.sort
end

def dfs(i)
  if @v[i]
    return false if 0 == @type[i]
    return true if 1 == @type[i] 
    return false if 2 == @type[i]
  end
  return true if 1 == @type[i]
  return false if 2 == @type[i]
  @v[i] = true
  is_safe = true
  @g[i].each do |j|
    is_safe = is_safe && dfs(j)
  end
  @type[i] = is_safe ? 1 : 2
  if is_safe
    @ans.push(i)
  end
  return is_safe
end

File: 20230712_Wednesday/20230712.go
// Leetcode: 802. Find Eventual Safe States.
// https://leetcode.com/problems/find-eventual-safe-states/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 105 ms, faster than 70.00% of Go online submissions for Find
// Eventual Safe States.
// Memory Usage: 7.6 MB, less than 41.43% of Go online submissions for Find
// Eventual Safe States.
// 2023.07.13 Updated.
package main

import (
	"sort"
)

var n int
var g [][]int
var ans []int
var types []int
var v []bool

func eventualSafeNodes(graph [][]int) []int {
	g = graph
	n = len(g)
	ans = make([]int, 0)
	// 0 - no info; 1 - safe; 2 - unsafe;
	types = make([]int, n)
	v = make([]bool, n)
	for i := 0; i < n; i += 1 {
		if !v[i] {
			dfs(i)
		}
	}
	sort.Ints(ans)
	return ans
}

func dfs(i int) bool {
	if v[i] {
		switch types[i] {
		case 0:
			return false
		case 1:
			return true
		case 2:
			return false
		}
	}
	switch types[i] {
	case 1:
		return true
	case 2:
		return false
	}
	v[i] = true
	isSafe := true
	for _, j := range g[i] {
		isSafe = isSafe && dfs(j)
	}
	if isSafe {
		types[i] = 1
		ans = append(ans, i)
	} else {
		types[i] = 2
	}
	return isSafe
}
